src/stac_task/task.py

```diff
@@ -0,0 +1,148 @@
+from abc import ABC, abstractmethod
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import Any, ClassVar, Dict, Generic, List, Optional, Type, TypeVar
+
+import pystac
+import pystac.utils
+import stac_asset.blocking
+from pydantic import BaseModel, Field
+
+from .models import Anything, Href, Item
+
+Input = TypeVar("Input", bound=BaseModel)
+Output = TypeVar("Output", bound=BaseModel)
+
+
+class Task(BaseModel, ABC, Generic[Input, Output]):
+    """A generic task."""
+
+    # Go away mypy, you can't handle this (it's not your fault,
+    # https://github.com/python/mypy/issues/5144)
+    input: ClassVar[Type[Input]] = Anything  # type: ignore
+    """The input model."""
+
+    output: ClassVar[Type[Output]] = Anything  # type: ignore
+    """The output model."""
+
+    payload_href: Optional[str] = None
+    """The href of the payload that was used to execute this task."""
+
+    working_directory: str = Field(default_factory=TemporaryDirectory)
+    """The directory to save any downloaded files."""
+
+    @abstractmethod
+    def process(self, item: Input) -> List[Output]:
```

Comment:
Handing in item (though should be ItemCollection) to the process function is way better than the current method which is to store the original items in the class.

I'm not sure if this needs to specify here that subclasses can define keywords, but a subclass can have keywords. The task parameters are then handed into the process function from those parameters.

Curious about other approaches but the advantage here is that it gives a task writer a simple experience - write the process function and any variables you need you add as keywords which will be automatically passed in via the payload.

Reply:
Keyword are passed on task instantiation, since they're constant for every call to `process`. A simple example from the tests: https://github.com/stac-utils/stac-task/blob/87bf2f02eb45c6d880a64f80bed76b018f31c4cb/tests/test_payload.py#L58-L63

You define the parameters for your task on the Task itself, which is a pydantic model: https://github.com/stac-utils/stac-task/blob/87bf2f02eb45c6d880a64f80bed76b018f31c4cb/tests/test_payload.py#L19-L28

Reply:
Ok, so the payload parameters are used instantiate the task and are available as class members?  This seems weird to me as an old-schooler, defining keywords seems simpler, but whatever the Pydantic way is is fine.

Reply:
Nope, they're instance members, added to `__init__` by pydantic: https://github.com/stac-utils/stac-task/blob/87bf2f02eb45c6d880a64f80bed76b018f31c4cb/src/stac_task/payload.py#L89

I'm 👎🏼 on `**kwargs` because it requires a developer to read code (or possibly out-of-date documentation) to know values are supported and typed as, and allowing subclasses to add new arguments to an overriden method violates https://en.wikipedia.org/wiki/Liskov_substitution_principle (and makes **mypy** cranky).